### PR TITLE
feat: add checksum support

### PR DESCRIPTION
**Pull request by MujkicA**
# Release notes

Added EIP-55 checksum address support by introducing `checksum_encode` and `is_checksum_valid` functions. Added the custom query `is_user_account` to `Provider` and `RetryableClient` to verify an address doesn't correspond to a different type of resource.

# Summary

The `checksum_encode` function produces checksummed addresses as strings, and `is_checksum_valid` verifies the correctness of a given address. 
The custom query `is_user_account` will check whether the given address corresponds to a blob, contract or transaction id.

# Checklist

- [x] All **changes** are **covered** by **tests**
- [ ] All **changes** are **documented** 
- [x] I **reviewed** the **entire PR** myself
- [x] I **described** all **Breaking Changes** (or there's none)

### CODE REVIEW

File: packages/fuels-core/src/types/checksum_address.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
+use crate::types::errors::{Error, Result};
+use sha3::{Digest, Keccak256};
+
+pub fn checksum_encode(address: &str) -> Result<String> {
+    let trimmed = address.trim_start_matches("0x");
+    pre_validate(trimmed)?;
+
+    let lowercase = trimmed.to_ascii_lowercase();
+
+    let hash = Keccak256::digest(lowercase.as_bytes());
```
Comment (mchristopher) [inline packages/fuels-core/src/types/checksum_address.rs:10]: We're not using Keccak on the TS side - can you confirm the test cases we have in https://github.com/FuelLabs/fuels-ts/pull/3308/files#diff-2c4d8f2d738f195a78c5ee15cb8b4271401b169fefbdccb0e1924b33e74b1cabR381-R400 all pass the same way here?

Comment (MujkicA) [reply packages/fuels-core/src/types/checksum_address.rs:10]: I'll switch it over to sha256 if that's what we want to do

File: Cargo.toml [hunk 2]
```diff
@@ -48,6 +48,8 @@ async-trait = { version = "0.1.74", default-features = false }
 bech32 = "0.9.1"
 bytes = { version = "1.5.0", default-features = false }
 chrono = "0.4.31"
+cynic = { version = "2.2", default-features = false }
+cynic-codegen = { version = "3.8.0", default-features = false }
```
Comment (segfault-magnet) [inline Cargo.toml:52]: Not used

File: packages/fuels-accounts/Cargo.toml [hunk 3]
```diff
@@ -33,6 +34,10 @@ fuel-tx = { workspace = true, features = ["test-helpers", "random"] }
 tempfile = { workspace = true }
 tokio = { workspace = true, features = ["test-util"] }
 
+[build-dependencies]
+cynic = { workspace = true, features = ["default"], optional = true } # TODO reduce features
```
Comment (segfault-magnet) [inline packages/fuels-accounts/Cargo.toml:38]: Let's maybe leave the default features. Buys us more fwd compat

File: packages/fuels-accounts/build.rs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+fn main() {
+    #[cfg(not(feature = "std"))]
```
Comment (segfault-magnet) [inline packages/fuels-accounts/build.rs:2]: Should run when std, so the opposite of this.

File: packages/fuels-accounts/src/provider/retryable_client.rs [hunk 5]
```diff
@@ -304,5 +306,55 @@ impl RetryableClient {
...
+        });
+
+        let response = self.client.query(query).await?;
+
+        let is_resource = response.blob.is_some()
+            || response.contract.is_some()
+            || response.transaction.is_some();
+
+        Ok(!is_resource)
+    }
```
Comment (segfault-magnet) [inline packages/fuels-accounts/src/provider/retryable_client.rs:328]: Move before delegation start

**Review by segfault-magnet — APPROVED** [review]

**Review by hal3e — APPROVED** [review]

**Review by digorithm — APPROVED** [review]

**Pull request merged**